lib/src/webcrypto/webcrypto.aesgcm.dart

```diff
@@ -14,40 +14,263 @@
+  /// _additional authenticated data_ (also called _associated data_) for
+  /// the encryption operation. Unlike the plaintext [data], the
+  /// [additionalData] is not encrypted. But integrity of the [additionalData]
+  /// is protected. Meaning that decryption will not succeed if [additionalData]
+  /// has be modified.
+  /// In an [authenticated encryption][2] scheme [additionalData] is typically
+  /// used to encode a IP, port, headers, date-time, a sequence number or
+  /// similar data that indicates how the ciphertext should be used.
+  /// As such [additionalData] aims to stop the ciphertext from being used
+  /// out of context.
+  ///
+  /// This operation requires a [tagLength], which specifies the bit-length
+  /// of the resulting authentication tag.
+  /// The permitted values for [tagLength] are:
+  ///  * `32` bits,
+  ///  * `64` bits,
+  ///  * `96` bits,
+  ///  * `104` bits,
+  ///  * `112` bits,
+  ///  * `120` bits, or,
+  ///  * `128` bits (default).
+  ///
+  /// This tag ensures the authenticity of the plaintext and the
+  /// [additionalData]. A short tag, may decrease these assurances.
+  /// For a discussion of [tagLength] and security assurances,
+  /// see [Appendix B of NIST SP 800-38D][1].
+  ///
+  /// This methods returns a [Uint8List] that is the concatenation of the
+  /// _ciphertext_ and the _authentication tag_.
+  /// That is, if you use a default [tagLength] of `128`, then the last 16 bytes
+  /// of the return value makes up the _authentication tag_.
+  ///
+  /// {@template AesGcmSecretKey-encryptBytes/decryptBytes:example}
+  /// **Example**
+  /// ```dart
+  /// import 'dart:convert' show utf8;
+  /// import 'dart:typed_data' show Uint8List;
+  /// import 'package:webcrypto/webcrypto.dart';
+  ///
+  /// // Generate a new random AES-GCM secret key for AES-256.
+  /// final k = await AesGcmSecretKey.generate(256);
+  ///
+  /// // Use a unique IV for each message.
+  /// final iv = Uint8List(16);
+  /// fillRandomBytes(iv);
+  ///
+  /// // Specify optional additionalData
+  /// final ad = utf8.encode('my-test-message');
+  ///
+  /// // Encrypt a message
+  /// final c = await k.encryptBytes(
+  ///   utf8.encode('hello world'),
+  ///   iv,
+  ///   additionalData: ad,
+  /// );
+  ///
+  /// // Decrypt message (requires the same iv)
+  /// print(utf8.decode(await k.decryptBytes(
+  ///   c,
+  ///   iv,
+  ///   additionalData: ad,
+  /// ))); // hello world
+  /// ```
+  /// {@endtemplate}
+  ///
+  /// {@template AesGcmSecretKey-remark:no-stream-api}
+  /// **Remark** this package does not offer a streaming API for
+  /// encryption / decryption using AES-GCM, because reading deciphered
+  /// plaintext prior to complete verification of the tag breaks the
+  /// authenticity assurances. Specifically, until the entire message is
+  /// decrypted it is not possible to know if it is authentic, which would
+  /// defeat the purpose of _authenticated encryption_.
```

Comment:
Theoretically there could be a streaming API for computational reasons (ie. you don't have to keep everything in memory at once), ending up with the result  being thrown away at the end.

Reply:
Technically correct, but BoringSSL doesn't offer a streaming interface, I suspect I read in a bug somewhere that it doesn't because the authors thought it would be a bad idea.

So that is the opinion we're carrying into `package:webcrypto`.

Also Web Cryptography specification doesn't offer streaming APIs at all, so I only added them because BoringSSL does. And on web we always buffer up the entire stream before we do the operation :rofl:

Reply:
I'd either remove the explanation or link to the discussion. I don't think we need to explain this.

I kind-of agree that a streaming interface would be hard to use correctly. But so is basically all crypto primitives. You really have to know what you are doing.